sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/RewriteDeleteFromTable.scala

```diff
@@ -0,0 +1,186 @@
+          DeleteFromTable(relation, cond, Some(rewritePlan))
+
+        case DataSourceV2Relation(_: SupportsDelete, _, _, _, _) =>
+          // don't assign a rewrite plan as the table supports deletes only with filters
+          delete
+
+        case DataSourceV2Relation(t, _, _, _, _) =>
+          throw new AnalysisException(s"Table $t does not support DELETE statements")
+
+        case _ =>
+          delete
+      }
+  }
+
+  // build a rewrite plan for sources that support replacing groups of data (e.g. files, partitions)
+  private def buildReplaceDataPlan(
+      relation: DataSourceV2Relation,
+      table: RowLevelOperationTable,
+      cond: Option[Expression]): LogicalPlan = {
+
+    // resolve all columns needed for DELETE including metadata columns for grouping data on write
+    val requiredWriteAttrs = resolveRefs(relation, table.operation.requiredWriteAttributes)
+
+    // construct a scan relation
+    val scanAttrs = toScanAttrs(relation, requiredWriteAttrs)
+    val scanRelation = relation.copy(table = table, output = scanAttrs)
+
+    // construct a plan that contains unmatched rows in matched groups that must be carried over
+    // such rows do not match the condition but have to be copied over as the source can replace
+    // only groups of rows
+    val deleteCond = cond.getOrElse(Literal.TrueLiteral)
+    val remainingRowsFilter = Not(EqualNullSafe(deleteCond, Literal(true, BooleanType)))
+    val remainingRowsPlan = Filter(remainingRowsFilter, scanRelation)
+
+    // pass only required columns for unmatched rows to the writer
+    val writeRelation = relation.copy(table = table, output = requiredWriteAttrs)
+    ReplaceData(writeRelation, remainingRowsPlan, relation)
+  }
+
+  // build a rewrite plan for sources that support row-level changes
+  private def buildWriteDeltaPlan(
+      relation: DataSourceV2Relation,
+      table: RowLevelOperationTable,
+      cond: Option[Expression]): LogicalPlan = {
+
+    val merge = table.operation.asInstanceOf[SupportsDelta]
+
+    // resolve all columns needed for merge (e.g. row ID and columns for grouping data on write)
+    val requiredWriteAttrs = resolveRefs(relation, table.operation.requiredWriteAttributes)
+    val rowIdAttrs = resolveRefs(relation, merge.rowId)
+
+    // construct a scan relation and include all required columns
+    val scanAttrs = toScanAttrs(relation, requiredWriteAttrs)
+    val scanRelation = relation.copy(table = table, output = scanAttrs)
+
+    // construct a plan that only contains records to delete
+    val deleteCond = cond.getOrElse(Literal.TrueLiteral)
+    val deletedRowsPlan = Filter(deleteCond, scanRelation)
+    val operation = Alias(Literal(DELETE_OPERATION), OPERATION_COLUMN)()
+    val project = Project(operation +: requiredWriteAttrs, deletedRowsPlan)
+
+    // pass operation and required columns to the writer as we are working with deltas
+    val operationAttr = resolveRef(project, FieldReference(OPERATION_COLUMN))
```

Comment:
This part is a bit ugly as we have an extra column that I currently make part of relation output so that the default name-based resolution works. An alternative is to add references to `WriteDelta` and have some custom resolution logic.